src/inmanta/ast/type.py

```diff
@@ -396,6 +421,13 @@ def __init__(self) -> None:
     def type_string(self) -> str:
         return "list"
 
+    def get_base_type(self) -> Type:
+        # The `list` type is not multi, thus it is the base type itself
+        return self
```

Comment:
We have to improve the definition of what multi means, as this is unexpected.

Reply:
I agree, I've spent a lot of time just thinking about this part. I didn't manage to figure out a cleaner approach. The two ways of declaring lists in the DSL makes this really hard. Suggestions are welcome.

Reply:
What happens is we

- add is_multi to the types
- make is_multi on the attribute use on the is_multi  of its type
- make this an is_multi type (so that is becomes the same as the regular list)

Reply:
I started out with the first two but reverted mostly because it doesn't work for `Entity`. There's no way to know from the type alone (currently) whether it's multi or not. Adding an `is_multi` method to `Type` would also add it to `Entity`. That said it's probably not required at this point either.
Your third point would make things a lot easier, but I'm not convinced it's correct. The current semantics of `is_multi` would then mean that `list` after this change is equivalent to `list[]` before the change.
Perhaps we could divide `Type` in internal helper types and those representing actual inmanta types. There would then be two ABC's, the second providing an abstract method `get_base_type()`. `TypedList` as well as `LiteralList` could then both implement this second ABC. This way we somewhat seperate the internal type hierarchy from the inmanta type hierarchy. I know this is still a bit vague, but I think we could find a pattern where we establish a seperate inmanta type hierarchy while still making use of the abstractions of the internal representations. This is something I just thought of, it's still very vague and I haven't given it sufficient thought to know if it's actually a good idea, but it might be worth exploring, though probably not as part of this issue.
For now, what do you think of my objection above?

Reply:
I would say that entity is not multi and list[] is actually wrong (we don't support string[][] nor list[] )

Reply:
Currently, we actually do support `list[]`. But even if we wouldn't, setting it to `multi` would result in lsm for example deriving `list[]`.

Reply:
We don't produce an error on list[], but is not used in any code nor any test and I would be surprised if it did work. To interpert it as list would be a step forward imho

Reply:
I strongly disagree on that last part. Not allowing it would be fine, but `list[]` clearly means a list of lists according to the semantics of `[]`, so interpreting it as a list of anything would be very counterintuitive.
But I think this is actually besides the point for this discussion. It's the other way around in this case. If we declare `List` to be multi with basetype `List`, that clearly impliest that it's a list of lists.

Reply:
We should raise an error when you make such a list of lists, it is an oversight that this was not the case already.

Reply:
but either way, I think this allows my three steps to be executed?

Reply:
Worse, imagine there's a validation type which is a list: `constrained_list`. The lsm module would take that type, and add `[]` to it because the attribute is multi. The validation type would then become `constrained_list[]` instead of `constrained_list`.